### PR TITLE
Generalized search_mesh that takes a functor to execute at element boundary or destination

### PR DESCRIPTION
**Pull request by Fuad-HH**
This new `particle_search` method allows passing a functor that takes all the necessary objects it wants to handle. It is run every time the particle reaches an element boundary or its destination i.e. after every call of `find_exit_face`. It will allow various things including handling different boundary conditions, recording information, etc.

### CODE REVIEW

**Comment by cwsmith** [comment]
Thank you.  Does this support 2d and/or 3d meshes?  Are there any tests available?

**Comment by Fuad-HH** [comment]
Yes, as the original `search_mesh` function, it supports both 2D and 3D. Actually, this is exactly the original `search_mesh` method except [what it did when particles reach element boundaries](https://github.com/SCOREC/pumi-pic/blob/87a7570f39fd2912b3b65938265edd901904f016/src/pumipic_adjacency.tpp#L489-L496) is encapsulated in [the functor](https://github.com/Fuad-HH/pumi-pic/blob/fc8dbc7a2adc86c6f3bbd9fce0817411e46ff182/src/pumipic_adjacency.tpp#L553-L572) that is passed in the new `particle_search` method. There are two test cases (search_adj_2d and search_adj_3d) that use the `search_mesh` method for 2D and 3D searches. It tests the `search_mesh` method and it is passing.

**Comment by jacobmerson** [comment]
@Fuad-HH is #128 ready as well? That has the unit tests you put in place for all of the component functions of `search_mesh` correct?

File: src/pumipic_adjacency.tpp [hunk 1]
```diff
@@ -544,5 +548,42 @@ namespace pumipic {
     Kokkos::Profiling::popRegion();
     return found;
   }
+
+  template <typename ParticleType, typename Segment3d>
+  struct NativeParticleHandlerAtElemBdry {
+    NativeParticleHandlerAtElemBdry () {}
```
Comment (jacobmerson) [inline src/pumipic_adjacency.tpp:554]: Based on above comments this functor will have some state data associated with it. I think the name can be something more descriptive about what this is actually doing. How about `RemoveParticleOnGeometricModelExit`?

File: src/pumipic_adjacency.tpp [hunk 2]
```diff
@@ -395,16 +395,25 @@ namespace pumipic {
...
-  template <class ParticleType, typename Segment3d, typename SegmentInt>
-  bool search_mesh(o::Mesh& mesh, ParticleStructure<ParticleType>* ptcls,
+
+  template <class ParticleType, typename Segment3d, typename SegmentInt, typename Func>
+  bool particle_search(o::Mesh& mesh, ParticleStructure<ParticleType>* ptcls,
                    Segment3d x_ps_orig, Segment3d x_ps_tgt, SegmentInt pids,
                    o::Write<o::LO>& elem_ids,
                    bool requireIntersection,
                    o::Write<o::LO>& inter_faces,
                    o::Write<o::Real>& inter_points,
```
Comment (jacobmerson) [inline src/pumipic_adjacency.tpp:405]: I think `inter_faces` and `inter_points` are specific for a specific choice of behavior at the element boundaries. These should be inside the functor, not in the function signature here.

File: src/pumipic_adjacency.tpp [hunk 3]
```diff
@@ -395,16 +395,25 @@ namespace pumipic {
...
   }
-  
-  template <class ParticleType, typename Segment3d, typename SegmentInt>
-  bool search_mesh(o::Mesh& mesh, ParticleStructure<ParticleType>* ptcls,
+
+  template <class ParticleType, typename Segment3d, typename SegmentInt, typename Func>
+  bool particle_search(o::Mesh& mesh, ParticleStructure<ParticleType>* ptcls,
                    Segment3d x_ps_orig, Segment3d x_ps_tgt, SegmentInt pids,
                    o::Write<o::LO>& elem_ids,
                    bool requireIntersection,
```
Comment (jacobmerson) [inline src/pumipic_adjacency.tpp:403]: `requireIntersection` is specific to a particular type of behavior at element boundaries and should be removed from the function signature and put inside the functor instead.

File: src/pumipic_adjacency.tpp [hunk 4]
```diff
@@ -544,5 +548,42 @@ namespace pumipic {
...
+                    Segment3d x_ps_orig,
+                    Segment3d x_ps_tgt) const {
+      const auto side_is_exposed = mark_exposed_sides(&mesh);
+      // Check if intersection face is exposed
+      check_model_intersection(mesh, ptcls, x_ps_orig, x_ps_tgt, elem_ids,
+                               ptcl_done, lastExit, side_is_exposed,
+                               requireIntersection, inter_faces);
+
+      // Move to next element
+      set_new_element(mesh, ptcls, elem_ids, ptcl_done, lastExit);
```
Comment (jacobmerson) [inline src/pumipic_adjacency.tpp:570]: Is there ever a case where we would not call `set_new_element`? I think this should actually be outside of the functor and inside the `particle_search` function.

File: src/pumipic_adjacency.tpp [hunk 5]
```diff
@@ -544,5 +548,42 @@ namespace pumipic {
...
+
+  template <typename ParticleType, typename Segment3d>
+  struct NativeParticleHandlerAtElemBdry {
+    NativeParticleHandlerAtElemBdry () {}
+
+    void operator()(o::Mesh& mesh, ParticleStructure<ParticleType>* ptcls,
+                    o::Write<o::LO>& elem_ids, o::Write<o::LO>& inter_faces,
+                    o::Write<o::LO>& lastExit, o::Write<o::Real>& inter_points,
+                    o::Write<o::LO>& ptcl_done,
+                    bool requireIntersection,
```
Comment (jacobmerson) [inline src/pumipic_adjacency.tpp:560]: `requireIntersection` should not be an input to `operator()`, but member data isntead.

File: src/pumipic_adjacency.tpp [hunk 6]
```diff
@@ -544,5 +548,42 @@ namespace pumipic {
...
     return found;
   }
+
+  template <typename ParticleType, typename Segment3d>
+  struct NativeParticleHandlerAtElemBdry {
+    NativeParticleHandlerAtElemBdry () {}
+
+    void operator()(o::Mesh& mesh, ParticleStructure<ParticleType>* ptcls,
+                    o::Write<o::LO>& elem_ids, o::Write<o::LO>& inter_faces,
+                    o::Write<o::LO>& lastExit, o::Write<o::Real>& inter_points,
```
Comment (jacobmerson) [inline src/pumipic_adjacency.tpp:558]: `inter_faces` and `inter_points` should be member data in the functor not in `operator()` since it is not data that the search itself needs, but only specific to what happens at element intersections.

File: src/pumipic_adjacency.tpp [hunk 7]
```diff
@@ -395,16 +395,25 @@ namespace pumipic {
     printf("Min area is: %.15f, Planned tol is %.15f\n", min_area, tol);
     return tol;
   }
-  
-  template <class ParticleType, typename Segment3d, typename SegmentInt>
-  bool search_mesh(o::Mesh& mesh, ParticleStructure<ParticleType>* ptcls,
+
+  template <class ParticleType, typename Segment3d, typename SegmentInt, typename Func>
+  bool particle_search(o::Mesh& mesh, ParticleStructure<ParticleType>* ptcls,
```
Comment (jacobmerson) [inline src/pumipic_adjacency.tpp:400]: @cwsmith @Angelyr any suggestions on the naming of this search method? Maybe something more descriptive like, `adjacency_based_particle_update`?

**Review by jacobmerson — CHANGES_REQUESTED** [review]
Thanks for putting this together. There are a few minor issues that need to be resolved to clean up the interfaces so we are not requiring implementation specific data to be passed around, but this is looking very good.

Thanks!

File: src/pumipic_adjacency.tpp (hunk 4 above)
Comment (Fuad-HH) [reply src/pumipic_adjacency.tpp:570]: I fully agree with this. I will remove it outside the functor.

File: src/pumipic_adjacency.tpp (hunk 7 above)
Comment (cwsmith) [reply src/pumipic_adjacency.tpp:400]: A few ideas... none are great:
`find_ptcl_bounding_element`
`update_ptcl_bounding_element`
`update_element_association_of_particles` 
`set_particle_to_element_association`

File: src/pumipic_adjacency.tpp (hunk 2 above)
Comment (Fuad-HH) [reply src/pumipic_adjacency.tpp:405]: These are coming from the search and are the faces and points where the particles intersect.

File: src/pumipic_adjacency.tpp (hunk 3 above)
Comment (Fuad-HH) [reply src/pumipic_adjacency.tpp:405]: The `requireIntersection` flag is used inside the function here. For example, `find_exit_face` needs it. Please find it [here](https://github.com/SCOREC/pumi-pic/blob/87a7570f39fd2912b3b65938265edd901904f016/src/pumipic_adjacency.tpp#L420)

Comment (jacobmerson) [reply src/pumipic_adjacency.tpp:405]: I see what you mean, but in tracking this down I did find something else we need to address to decide what to do here.

In this code
https://github.com/SCOREC/pumi-pic/blob/87a7570f39fd2912b3b65938265edd901904f016/src/pumipic_adjacency.tpp#L490-L493

We fill up `lastExit` and `inter_points` arrays. Where, these correspond to the element face that the particle passes through using `find_exit_face`.

However, in `check_model_intersection` we fill in the `inter_faces` array. I *think* (but could be wrong) that `inter_faces` is supposed to be the faces that the particle passes through on the boundary.

If this is correct, then what's currently called `inter_faces` should not be in the search, but should be inside the functor.

And, the naming of `lastExit` and `inter_points` needs to be made consistent. The current naming is confusing since it seems like `inter_points` and `inter_faces` are referring to the same intersection but those are not matched.

Comment (Fuad-HH) [reply src/pumipic_adjacency.tpp:405]: Yes, you are right. It should be in the functor. Search doesn't depend on it. You are correct about the naming. I had to write down what they do or how they are used.

File: src/pumipic_adjacency.tpp (hunk 7 above)
Comment (jacobmerson) [reply src/pumipic_adjacency.tpp:402]: I think this is the last thing to resolve before we can merge. How about `trace_particle_through_mesh` ?

Comment (cwsmith) [reply src/pumipic_adjacency.tpp:402]: That works.  We should put some doxygen style comments on the function and explicitly describe what it does.

**Comment by Fuad-HH** [comment]
I have added a brief description for renamed `trace_particle_through_mesh`.

**Comment by Fuad-HH** [comment]
There shouldn't be any merge conflict, but it's showing an "empty" line coming from `master`.

**Review by cwsmith — APPROVED** [review]
Looks good.  Thank you.

**Review by jacobmerson — APPROVED** [review]
